File: src/data/static/ExerciseNormalizer.ts
import type { Exercise, MuscleInvolvement } from '../../domain/models/Exercise'
import type { RawExercise, SourceMuscleName } from './source/sourceSchema'
import { MuscleRole } from '../../domain/enums/MuscleRole'
import {
  SOURCE_MUSCLE_TO_ID,
  SOURCE_EQUIPMENT_TO_ENUM,
  SOURCE_MECHANIC_TO_ENUM,
  SOURCE_FORCE_TO_ENUM,
  SOURCE_LEVEL_TO_ENUM,
  SOURCE_CATEGORY_TO_ENUM,
  ROLE_DEFAULT_CONTRIBUTION,
} from './mapping/sourceMuscleMap'

/**
 * Transforms raw `free-exercise-db` records into our immutable `Exercise`
 * entities: maps the source vocabulary to our enums/taxonomy, derives
 * muscle involvements (primary/secondary) with default contributions, and
 * resolves relative image paths against the configured CDN base.
 *
 * Stabilizer involvements are intentionally not invented here — the source
 * has none, and fabricating them would be guesswork. They are a later
 * curation pass (see DATA_MODEL.md).
 */
export class ExerciseNormalizer {
  constructor(private readonly imageBaseUrl: string) {}

  normalizeAll(raws: readonly RawExercise[]): readonly Exercise[] {
    return raws.map((raw) => this.normalize(raw))
  }

  normalize(raw: RawExercise): Exercise {
    return {
      id: raw.id,
      name: raw.name,
      muscles: this.toInvolvements(raw),
      category: SOURCE_CATEGORY_TO_ENUM[raw.category],
      level: SOURCE_LEVEL_TO_ENUM[raw.level],
      equipment: raw.equipment ? SOURCE_EQUIPMENT_TO_ENUM[raw.equipment] : undefined,
      mechanic: raw.mechanic ? SOURCE_MECHANIC_TO_ENUM[raw.mechanic] : undefined,
      force: raw.force ? SOURCE_FORCE_TO_ENUM[raw.force] : undefined,
      instructions: raw.instructions,
      images: raw.images.map((path) => this.resolveImageUrl(path)),
    }
  }

  private toInvolvements(raw: RawExercise): readonly MuscleInvolvement[] {
    return [
      ...raw.primaryMuscles.map((m) => this.involvement(m, MuscleRole.Primary)),
      ...raw.secondaryMuscles.map((m) => this.involvement(m, MuscleRole.Secondary)),
    ]
  }

  private involvement(source: SourceMuscleName, role: MuscleRole): MuscleInvolvement {
    return {
      muscleId: SOURCE_MUSCLE_TO_ID[source],
      role,
      contribution: ROLE_DEFAULT_CONTRIBUTION[role],
    }
  }

  private resolveImageUrl(path: string): string {
    return `${this.imageBaseUrl}${path}`
  }
}

File: backend/src/main/java/com/musclemap/catalog/ExerciseNormalizer.java
package com.musclemap.catalog;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Transforms raw free-exercise-db records into {@link Exercise} entities: maps
 * the source vocabulary onto our enums/taxonomy, derives muscle involvements
 * (primary/secondary) with default contributions, resolves image paths against
 * the CDN, and prepends a curated YouTube demo video when one exists. Mirrors the
 * frontend {@code ExerciseNormalizer} so the API and the static fallback agree.
 *
 * <p>Unknown source values fail fast (an {@link IllegalStateException}) rather
 * than silently producing a NULL — that surfaces dataset drift at boot.</p>
 */
class ExerciseNormalizer {

    private final CatalogProperties properties;
    private final Map<String, String> videoIds;

    ExerciseNormalizer(CatalogProperties properties, Map<String, String> videoIds) {
        this.properties = properties;
        this.videoIds = videoIds;
    }

    Exercise normalize(RawExercise raw) {
        Exercise exercise = new Exercise();
        exercise.setId(raw.id());
        exercise.setName(raw.name());
        exercise.setCategory(require(CatalogTaxonomy.SOURCE_CATEGORY, raw.category(), "category"));
        exercise.setLevel(require(CatalogTaxonomy.SOURCE_LEVEL, raw.level(), "level"));
        exercise.setEquipment(optional(CatalogTaxonomy.SOURCE_EQUIPMENT, raw.equipment(), "equipment"));
        exercise.setMechanic(optional(CatalogTaxonomy.SOURCE_MECHANIC, raw.mechanic(), "mechanic"));
        exercise.setForce(optional(CatalogTaxonomy.SOURCE_FORCE, raw.force(), "force"));
        exercise.setInstructions(new ArrayList<>(raw.instructionsOrEmpty()));
        exercise.setMuscles(toInvolvements(raw));
        exercise.setMedia(toMedia(raw));
        return exercise;
    }

    private List<MuscleInvolvement> toInvolvements(RawExercise raw) {
        List<MuscleInvolvement> involvements = new ArrayList<>();
        for (String source : raw.primaryMusclesOrEmpty()) {
            involvements.add(involvement(source, MuscleRole.PRIMARY));
        }
        for (String source : raw.secondaryMusclesOrEmpty()) {
            involvements.add(involvement(source, MuscleRole.SECONDARY));
        }
        return involvements;
    }

    private MuscleInvolvement involvement(String sourceMuscle, MuscleRole role) {
        String muscleId = CatalogTaxonomy.SOURCE_MUSCLE_TO_ID.get(sourceMuscle);
        if (muscleId == null) {
            throw new IllegalStateException("Unknown source muscle: " + sourceMuscle);
        }
        return new MuscleInvolvement(muscleId, role, CatalogTaxonomy.ROLE_DEFAULT_CONTRIBUTION.get(role));
    }

    /** A curated demo video (if any) first, then the source images. */
    private List<ExerciseMedia> toMedia(RawExercise raw) {
        List<ExerciseMedia> media = new ArrayList<>();
        String videoId = videoIds.get(raw.id());
        if (videoId != null && !videoId.isBlank()) {
            media.add(new ExerciseMedia(MediaKind.VIDEO, MediaSource.YOUTUBE, videoId,
                    properties.youtubeThumbnailUrl(videoId)));
        }
        for (String path : raw.imagesOrEmpty()) {
            media.add(new ExerciseMedia(MediaKind.IMAGE, MediaSource.FILE, properties.imageUrl(path), null));
        }
        return media;
    }

    private static <T> T require(Map<String, T> table, String key, String field) {
        T value = table.get(key);
        if (value == null) {
            throw new IllegalStateException("Unknown exercise " + field + ": " + key);
        }
        return value;
    }

    private static <T> T optional(Map<String, T> table, String key, String field) {
        if (key == null) {
            return null;
        }
        return require(table, key, field);
    }
}
